### DISCUSSION TITLE
How is the 2FA flow supposed to work when logging in?

### DISCUSSION THREAD

**Discussion by joekrill** [ory/kratos#2996] [Q&A]
What is the expected behavior for the 2FA flow? I'm currently trying to get it to work. I initialize a login flow and login successfully with a username and password, which works fine. But this returns a 200 success response with a new `session` object! Except `session.identity` is `null`. It's not until i do a `/sessions/whoami` request that I get a 403 response with the error id of `session_aal2_required`. 

I was expecting that the original login flow POST with a valid username/password would return something indicating that there is a further action needed (aal2 in this case), and perhaps update the flow `ui` with the totp nodes. 

The current  behavior is a little bizarre because from the application perspective, the user seems to have logged in, so I redirect them back to wherever they were before they clicked login. But then I have to just redirect them to the aal2 flow. I have it working where I just show the aal2 login form in a modal anytime `/sessions/whoami` returns the `session_aal2_required`, but it feels very clunky. 

Is this still a work in progress? Or am I misunderstanding how this is supposed to work? Is this maybe the same bug that is mentioned here?:  https://github.com/ory/kratos/discussions/2742

**Comment by davedx** [comment]
I managed to get reasonably far implementing MFA, but it's been hard going due to an almost total lack of documentation for how the front-end is supposed to work. Basically if you follow the docs on how to configure the back-end, that is correct and does work. Then on the front-end, you need to first present the user with the UI settings nodes that are in the totp group (if you're using that). This will allow them to add an authenticator to their profile.

You should then get ui nodes in the login page that you can render for aal2 (enter authentication code), along with messages for the user. Submitting this flow to the kratos client will successfully upgrade their session to aal2 level.

Where I got blocked was that I removed my temporary authenticator identity from my phone, my user account was then stuck. I wanted to implement a way for us to handle this in our application, because it's a very real use case, but started running into weird issues like the whoami endpoint returned 200, but the settings endpoint returned 403, so the user could no longer update their MFA settings.

I think that basically MFA in Kratos self service does work, but the documentation is almost non existent, and given this is security I'm quite wary of proceeding with my implementation because I just don't know if I'm doing it right. (I also don't know how complete the implementation is on the Kratos side).

**Comment by Benehiko** [comment]
Hi @joekrill and @davedx 

I'm sorry to hear you've been struggling with this. We try our best to provide as much documentation as possible to make the process a lot easier on you. Unfortunately, 2FA is indeed a complicated topic and depending on what you would like to support, e.g. hardware keys or totp codes the process of implementation is different.  

Please take a look at these documents:
- https://www.ory.sh/docs/kratos/mfa/overview
- https://www.ory.sh/docs/kratos/bring-your-own-ui/custom-ui-advanced-integration#two-factor-authentication

I would suggest trying this out through the [Account Experience](https://console.ory.sh/) (the managed UI bundled with each Ory Network project) to get a feel for how it works. We have also created a component library ([Ory Elements](https://github.com/ory/elements)) with examples in NextJs, React and Preact which implements all of the flows, including MFA. I think this is a great place to get started when implementing your custom UI.

> I also don't know how complete the implementation is on the Kratos side

Kratos has been released out of alpha for quite some time now ([since May 2022](https://github.com/ory/kratos/releases/tag/v0.10.0)) and is being used in the [Ory Network](https://console.ory.sh/). We are also constantly bug fixing and shipping new features.
